src/more_ops.rs

```diff
@@ -24,56 +24,74 @@ use sha2::{Digest, Sha256};
 #[cfg(unix)]
 use openssl::sha;
 
-const ARITH_BASE_COST: Cost = 1;
-const ARITH_COST_PER_ARG: Cost = 3;
-const ARITH_COST_PER_BYTE_DIVIDER: Cost = 16;
+const ARITH_BASE_COST: Cost = 99;
+const ARITH_COST_PER_ARG: Cost = 320;
+const ARITH_COST_PER_BYTE: Cost = 3;
 
-const LOG_BASE_COST: Cost = 1;
-const LOG_COST_PER_ARG: Cost = 3;
-const LOG_COST_PER_BYTE_DIVIDER: Cost = 16;
+const LOG_BASE_COST: Cost = 100;
+const LOG_COST_PER_ARG: Cost = 310;
+// increased from 3 to charge more for memory allocations
+const LOG_COST_PER_BYTE: Cost = 6;
 
-const LOGNOT_BASE_COST: Cost = 3;
-const LOGNOT_COST_PER_BYTE_DIVIDER: Cost = 16;
+const LOGNOT_BASE_COST: Cost = 331;
+// increased from 3 to charge more for memory allocations
+const LOGNOT_COST_PER_BYTE: Cost = 6;
 
-const MUL_BASE_COST: Cost = 1;
-const MUL_COST_PER_OP: Cost = 8;
-const MUL_LINEAR_COST_PER_BYTE_DIVIDER: Cost = 16;
-const MUL_SQUARE_COST_PER_BYTE_DIVIDER: Cost = 16384;
+const MUL_BASE_COST: Cost = 92;
+const MUL_COST_PER_OP: Cost = 885;
+const MUL_LINEAR_COST_PER_BYTE: Cost = 6;
+const MUL_SQUARE_COST_PER_BYTE_DIVIDER: Cost = 128;
 
-const GR_BASE_COST: Cost = 5;
-const GR_COST_PER_BYTE_DIVIDER: Cost = 32;
+const GR_BASE_COST: Cost = 590;
+const GR_COST_PER_BYTE: Cost = 2;
 
-const CMP_BASE_COST: Cost = 2;
-const CMP_COST_PER_BYTE_DIVIDER: Cost = 64;
+const GRS_BASE_COST: Cost = 209;
+const GRS_COST_PER_BYTE: Cost = 1;
 
-const STRLEN_BASE_COST: Cost = 2;
-const STRLEN_COST_PER_BYTE_DIVIDER: Cost = 128;
+const STRLEN_BASE_COST: Cost = 219;
+const STRLEN_COST_PER_BYTE: Cost = 1;
 
-const CONCAT_BASE_COST: Cost = 1;
-const CONCAT_COST_PER_ARG: Cost = 1;
-const CONCAT_COST_PER_BYTE_DIVIDER: Cost = 16;
+const CONCAT_BASE_COST: Cost = 142;
+const CONCAT_COST_PER_ARG: Cost = 135;
+// increased from 3 to charge more for memory allocations
+const CONCAT_COST_PER_BYTE: Cost = 6;
```

Comment:
I'd recommend that costs that are artificially cranked higher to prevent memory DOS attacks have a name with `ALLOCATED` or something in them, and maybe even a constant multiplier with like `ANTI_DOS_MALLOC_MULTIPLIER = 2`. Something like

```
const ANTI_DOS_MALLOC_MULTIPLIER: u32 = 2;
(...snip...)
const CONCAT_COST_PER_MALLOC_BYTE: Cost = 3 * ANTI_DOS_MALLOC_MULTIPLIER;

const MUL_LINEAR_COST_PER_MALLOC_BYTE: Cost = 3 * ANTI_DOS_MALLOC_MULTIPLIER;
```
This way we can crank up or down all anti-memory DOS vectors at once.

Reply:
that's a good point. Although, we won't be changing this for that many more days, but it makes for more understandable code

Reply:
True, we could delay that change until later as long as the constants' values don't change.

Reply:
I don't think the malloc cost should be a multiplier though. If it's meant to charge cost for allocating memory, it doesn't really matter how expensive the operation was to *produce* the byte, we charge for the allocation of the byte.

Reply:
part of me feels like this really should be a cost applied to `new_atom()`, based on the number of bytes. Right now, `lognot` allocates memory (as much as its input atom), but it pays less of the "malloc" cost compared to `logxor`, since the latter is binary and pays the malloc cost for both its operands.

It might be a bit late to make a change like this though.